File: transcription/whisper.go
//go:build whisper

package transcription

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	whisper "github.com/ggerganov/whisper.cpp/bindings/go/pkg/whisper"
)

// WhisperClient is a local batch STT adapter using whisper.cpp CGo bindings.
// Audio is buffered during recording; transcription runs when Close() is called.
//
// Model path resolution for STT_MODEL=whisper/<name>:
//   - "base"  → ~/.yogurt/whisper/ggml-base.bin
//   - "tiny"  → ~/.yogurt/whisper/ggml-tiny.bin
//   - "path:/absolute/path.bin" → explicit path
type WhisperClient struct {
	sampleRate int
	modelPath  string

	mu       sync.Mutex
	audioBuf []byte
	closed   bool

	onSegment    func(Segment)
	onError      func(error)
	onConnected  func()
	onDisconnect func()
}

func NewWhisperClient(sampleRate int, model string) STTClient {
	return &WhisperClient{
		sampleRate: sampleRate,
		modelPath:  resolveWhisperModel(model),
	}
}

func (c *WhisperClient) SetOnSegment(fn func(Segment))    { c.onSegment = fn }
func (c *WhisperClient) SetOnError(fn func(error))         { c.onError = fn }
func (c *WhisperClient) SetOnConnected(fn func())           { c.onConnected = fn }
func (c *WhisperClient) SetOnDisconnect(fn func())          { c.onDisconnect = fn }

func (c *WhisperClient) Connect() error {
	if _, err := os.Stat(c.modelPath); err != nil {
		return fmt.Errorf("whisper model not found at %s — download a model to that path", c.modelPath)
	}
	if c.onConnected != nil {
		go c.onConnected()
	}
	return nil
}

func (c *WhisperClient) SendAudio(data []byte) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.audioBuf = append(c.audioBuf, data...)
	return nil
}

func (c *WhisperClient) IsConnected() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return !c.closed
}

func (c *WhisperClient) Close() error {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil
	}
	c.closed = true
	buf := c.audioBuf
	c.mu.Unlock()

	defer func() {
		if c.onDisconnect != nil {
			c.onDisconnect()
		}
	}()

	if len(buf) == 0 {
		return nil
	}

	segments, err := c.transcribe(buf)
	if err != nil {
		if c.onError != nil {
			c.onError(err)
		}
		return err
	}
	for _, seg := range segments {
		if c.onSegment != nil {
			c.onSegment(seg)
		}
	}
	return nil
}

func (c *WhisperClient) transcribe(pcm []byte) ([]Segment, error) {
	model, err := whisper.New(c.modelPath)
	if err != nil {
		return nil, fmt.Errorf("load whisper model: %w", err)
	}
	defer model.Close()

	ctx, err := model.NewContext()
	if err != nil {
		return nil, fmt.Errorf("create whisper context: %w", err)
	}

	samples := pcm16ToFloat32(pcm)
	if err := ctx.Process(samples, nil, nil); err != nil {
		return nil, fmt.Errorf("whisper process: %w", err)
	}

	now := time.Now()
	var segs []Segment
	for {
		s, err := ctx.NextSegment()
		if err != nil {
			break
		}
		segs = append(segs, Segment{
			Text:      strings.TrimSpace(s.Text),
			Speaker:   "A",
			StartTime: s.Start.Seconds(),
			EndTime:   s.End.Seconds(),
			IsFinal:   true,
			CreatedAt: now,
		})
	}
	return segs, nil
}

// resolveWhisperModel converts a model name to an absolute path.
func resolveWhisperModel(model string) string {
	if strings.HasPrefix(model, "path:") {
		return strings.TrimPrefix(model, "path:")
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".yogurt", "whisper", "ggml-"+model+".bin")
}

// pcm16ToFloat32 converts raw PCM16 LE bytes to float32 samples.
func pcm16ToFloat32(pcm []byte) []float32 {
	samples := make([]float32, len(pcm)/2)
	for i := range samples {
		s := int16(pcm[2*i]) | int16(pcm[2*i+1])<<8
		samples[i] = float32(s) / 32768.0
	}
	return samples
}

File: src/stt/whisper.rs
//! Local batch STT via whisper.cpp (whisper-rs). Feature-gated: `--features whisper`.

use std::path::PathBuf;
use std::sync::Mutex;

use anyhow::{Result, bail};
use chrono::Local;

use super::{Segment, SttCallbacks, SttClient};

pub struct WhisperClient {
    sample_rate: u32,
    model_path: PathBuf,
    callbacks: SttCallbacks,
    audio_buf: Mutex<Vec<u8>>,
    closed: Mutex<bool>,
}

impl WhisperClient {
    pub fn new(sample_rate: u32, model: &str, callbacks: SttCallbacks) -> Self {
        WhisperClient {
            sample_rate,
            model_path: resolve_model(model),
            callbacks,
            audio_buf: Mutex::new(Vec::new()),
            closed: Mutex::new(false),
        }
    }

    fn transcribe(&self, pcm: &[u8]) -> Result<Vec<Segment>> {
        use whisper_rs::{FullParams, SamplingStrategy, WhisperContext, WhisperContextParameters};

        let samples: Vec<f32> = pcm
            .chunks_exact(2)
            .map(|c| i16::from_le_bytes([c[0], c[1]]) as f32 / 32768.0)
            .collect();
        log::info!(
            "whisper: transcribing {:.1}s of audio with {}",
            samples.len() as f64 / self.sample_rate as f64,
            self.model_path.display()
        );

        // whisper.cpp and GGML log to stderr by default, which corrupts the
        // alt-screen TUI. Route them through `log` (i.e. into yogurt.log).
        static LOG_HOOKS: std::sync::Once = std::sync::Once::new();
        LOG_HOOKS.call_once(whisper_rs::install_logging_hooks);

        let ctx = WhisperContext::new_with_params(
            &self.model_path.to_string_lossy(),
            WhisperContextParameters::default(),
        )?;
        let mut state = ctx.create_state()?;
        let mut params = FullParams::new(SamplingStrategy::Greedy { best_of: 1 });

        params.set_language(Some("auto"));
        state.full(params, &samples)?;

        let mut segments = Vec::new();
        for i in 0..state.full_n_segments()? {
            let text = state.full_get_segment_text(i)?.trim().to_string();
            if text.is_empty() {
                continue;
            }
            segments.push(Segment {
                text,
                speaker: "A".into(), // no diarization
                start_time: state.full_get_segment_t0(i)? as f64 * 0.01,
                end_time: state.full_get_segment_t1(i)? as f64 * 0.01,
                confidence: 0.0,
                is_final: true,
                created_at: Local::now(),
            });
        }
        Ok(segments)
    }
}

impl SttClient for WhisperClient {
    fn connect(&mut self) -> Result<()> {
        if !self.model_path.exists() {
            bail!(
                "whisper model not found at {} — run: make setup MODEL=<name>",
                self.model_path.display()
            );
        }
        if self.sample_rate != 16000 {
            bail!("whisper requires 16kHz audio (got {}Hz)", self.sample_rate);
        }
        (self.callbacks.on_connected)();
        Ok(())
    }

    fn send_audio(&self, pcm: &[u8]) -> Result<()> {
        self.audio_buf.lock().unwrap().extend_from_slice(pcm);
        Ok(())
    }

    fn close(&mut self) -> Result<()> {
        {
            let mut closed = self.closed.lock().unwrap();
            if *closed {
                return Ok(());
            }
            *closed = true;
        }
        let buf = std::mem::take(&mut *self.audio_buf.lock().unwrap());

        let result = (|| -> Result<()> {
            if buf.is_empty() {
                return Ok(());
            }
            for seg in self.transcribe(&buf)? {
                (self.callbacks.on_segment)(seg);
            }
            Ok(())
        })();

        if let Err(e) = &result {
            (self.callbacks.on_error)(e.to_string());
        }
        (self.callbacks.on_disconnect)();
        result
    }
}

/// "path:/abs/model.bin" → literal path; otherwise ~/.yogurt/whisper/ggml-<model>.bin
fn resolve_model(model: &str) -> PathBuf {
    if let Some(p) = model.strip_prefix("path:") {
        return PathBuf::from(p);
    }
    let name = if model.is_empty() { "base" } else { model };
    dirs::home_dir()
        .unwrap_or_default()
        .join(".yogurt")
        .join("whisper")
        .join(format!("ggml-{name}.bin"))
}
